### ISSUE TITLE
[FR]: sshd: kex_exchange_identification silently ignored (not banned)

### ISSUE THREAD

**Issue by allanwind** [fail2ban/fail2ban#3086]
### Environment:

- Fail2Ban version  : 0.11.2-2
- OS, including release name/version : Debian 11 (bullseye)

#### Service, project or product which log or journal should be monitored

- Name of filter or jail in Fail2Ban (if already exists) : sshd
- Service, project or product name, including release name/version : oopenssh-server 1:8.4p1-5
- Repository or URL (if known) : 
- Service type : 
- Ports and protocols the service is listening : 

#### Log or journal information

- Log file name(s) : 

- Journal identifier or unit name : 

#### Any additional information

1. The first two event are silently ignored but should trigger a ban.  The issue, I think, is while the first event matched via:
```
mdre-ddos = ...
            ^kex_exchange_identification: (?:[Cc]lient sent invalid protocol identifier|[Cc]onnection closed by remote host)
```

the 2nd event does not but I think was meant to:

```
mdre-ddos-other = ^<F-MLFFORGET>(Connection (?:closed|reset)|Disconnected)</F-MLFFORGET> (?:by|from)%(__authng_user)s <HOST>%(__on_port_opt)s\s+\[preauth\]\s*$
```

I don't know how to fix this, in particular, if it's safe to make those extra fields optional.

The will trigger a ban if you change sshd_config's LogLevel to VERBOSE but I think it's pretty bad that this just silently fails to do anything by default.

2. The last two events should also trigger a ban but doesn't.  This is probably ust a minor tweak of the regex in mdre-ddos to include the `read: Connection reset by peer` case.
3. There is a comment in sshd.conf:
```
"Connection from <HOST> port \d+" requires LogLevel VERBOSE in sshd_config`
```
but it was not helpful to me as it talks about the solution instead of the problem.  Maybe it should say something like this instead: any filter in failregex that doesnt' have a <HOST> will be silently ignored if no other event matches the session (prefregerx) unless you set sshd_config LogLevel VERBOSE (or higher?).

#### failures in sense of fail2ban filter (fail2ban must match):

```
2021-08-16T04:10:35.924+00:00 pawan sshd[424228]: error: kex_exchange_identification: Connection closed by remote host                                                                                                                         
2021-08-16T04:10:35.924+00:00 pawan sshd[424228]: Connection closed by 205.185.113.128 port 35352                                                                                                                                              
... and after setting LogLevel to VERBOSE ...
2021-08-17T04:29:10.174+00:00 pawan sshd[649921]: Connection from 46.37.81.241 port 51280 on 71.126.254.78 port 22 rdomain ""
2021-08-17T04:29:10.174+00:00 pawan sshd[649921]: error: kex_exchange_identification: read: Connection reset by peer
2021-08-17T04:29:10.174+00:00 pawan sshd[649921]: Connection reset by 46.37.81.241 port 51280

```

#### legitimate messages (fail2ban should not consider as failures):

```
```

**Comment by sebres** [comment]
Hmm, indirectly f77398c49d4eeb529a1684a27dcfbf5b6aaafa66 is probably to blame here.
This was initiated to match `Connection reset/close` on preauth phase as a failure for `ddos`/`aggressive` mode, but removed a helper (to retrieve IP) for same message without to consider `[preauth]`.

> 

Well, basically it is very simple to fix it:

```diff
 mdre-ddos-other = ^<F-MLFFORGET>(Connection (?:closed|reset)|Disconnected)</F-MLFFORGET> (?:by|from)%(__authng_user)s <HOST>%(__on_port_opt)s\s+\[preauth\]\s*$
+                  <mdre-normal-other>

```
so it will restore additional RE (`<F-NOFAIL>` helper) matching also closed/reset message without `[preauth]` as not a failure (only to obtain IP).

<details><summary>PoC ...</summary>

```
$ fail2ban-regex /tmp/log 'sshd[mode=aggressive]'

...

Results
=======

Prefregex: 2 total
|  ^(?P<mlfid>(?:\[\])?\s*(?:<[^.]+\.[^.]+>\s+)?(?:\S+\s+)?(?:kernel:\s?\[ *\d+\.\d+\]:?\s+)?(?:@vserver_\S+\s+)?(?:(?:(?:\[\d+\])?:\s+[\[\(]?sshd(?:\(\S+\))?[\]\)]?:?|[\[\(]?sshd(?:\(\S+\))?[\]\)]?:?(?:\[\d+\])?:?)\s+)?(?:\[ID \d+ \S+\]\s+)?)(?:(?:error|fatal): (?:PAM: )?)?(?P<content>.+)$
`-

Failregex: 2 total
|-  #) [# of hits] regular expression
|  21) [1] ^<F-NOFAIL><F-MLFFORGET>(Connection closed|Disconnected)</F-MLFFORGET></F-NOFAIL> (?:by|from)(?: (?:invalid|authenticating) user <F-USER>\S+|.*?</F-USER>)? <HOST>(?:(?: (?:port \d+|on \S+|\[preauth\])){0,3}\s*|\s*)$
|  24) [1] ^kex_exchange_identification: (?:[Cc]lient sent invalid protocol identifier|[Cc]onnection closed by remote host)
`-

...

Lines: 2 lines, 0 ignored, 2 matched, 0 missed


$ fail2ban-regex -o '<ip><br><matches>' /tmp/log 'sshd[mode=aggressive]'
192.0.2.1
2021-08-16T04:10:35.924+00:00 srv sshd[424228]: error: kex_exchange_identification: Connection closed by remote host
2021-08-16T04:10:35.924+00:00 srv sshd[424228]: Connection closed by 192.0.2.1 port 35352
```
</details>

(of course it will match in fail2ban server only if `mode = aggressive` or `mode = ddos` is specified in sshd jail).

**Comment by allanwind** [comment]
On 2021-08-17T05:11:22, Sergey G. Brester wrote:
>Well, basically it is very simple to fix it:
>
>```diff
> mdre-ddos-other = ^<F-MLFFORGET>(Connection (?:closed|reset)|Disconnected)</F-MLFFORGET> (?:by|from)%(__authng_user)s <HOST>%(__on_port_opt)s\s+\[preauth\]\s*$
>+                  <mdre-normal-other>
>
>```

Thank you Sergey for the context and the quick resolution.  This appears to be the relevant .local change:

[Definition]
mdre-ddos =
   %(known/mdre-ddos)s
   ^kex_exchange_identification: read: Connection reset by peer$
mdre-ddos-other =
   %(known/mdre-ddos-other)s
   ^<F-MLFFORGET>Connection reset</F-MLFFORGET> by <HOST>(?:%(known/__suff)s|\s*)$
   <mdre-normal-other>

Do you think it's safe to have mdre-ddos-other match on 
"Connection reset" like this?  Or any other suggestion how to 
accomplish the 2nd part?

**Comment by sebres** [comment]
You don't need to use known-prefix e. g. `%(known/__suff)s` for values you don't overwrite by yourself.
And it can be considered as safe at least as long as the substituted variables (like `__suff`) are available in original filter (and changing backwards compatible in possible upgrades).

But so regarding the all "current" versions of fail2ban. 
Newest version (v.1.0) gets a new enhancement which would reorder the regexp's by those matched count, so one would need to rewrite the RE to avoid matching `[preauth]` in suffix or to prohibit rising up for this RE (not so optimal), because at the moment same RE (with `[preauth]`) matches as failure in `ddos` mode, which can "conflict" with a new `NO-FAIL` helper from my interim suggestion, if it gets moved.

I'll try to provide new version of sshd-filter referencing this issue, where it would be correct, so you'd simply use this new version hereafter.

**Comment by allanwind** [comment]
On 2021-08-17T11:08:23, Sergey G. Brester wrote:
>You don't need to use known-prefix e. g. `%(known/__suff)s` for 
>values you don't overwrite by yourself.
>And it can be considered as safe at least as long as the 
>substituted variables (like `__suff`) are available in original 
>filter (and changing backwards compatible in possible upgrades).

Thanks.

Re, backwards compatibility, 2 (apache-noscript, sshd) of the 3  
.local filters silently "broke" with the upgrade from Debian 10 to 
11 as those jails added or changed prefregex which in turn meant
that my .local failregex had to change.  Just wanted to share the 
data point.  Both .local and prefregex are good features making 
both the upgrade itself seamless and the filters easier to read
(besides the performance improvement).

>Newest version (v.1.0) gets a new enhancement which would reorder 
>the regexp's by those matched count,

I appreciate the performance benefit.  It seems risky to mess with 
user input.  Hopefully there is an option to disable the feature.  
Maybe add debug option that hits all regex and log if multiple 
filters match as those would be the ones that could potentially 
change meaning when reordered?

**Comment by sebres** [comment]
> ... could potentially change meaning when reordered?

I can hardly imagine such a filters (or rather the needs were it must be handled intentionally with different meaning for same line).

> It seems risky to mess with user input.

What are the misgivings you have about this approach?
Anyway, it works on my side several years without any issue. 
The performance gain (and related reduction of fail2ban's load) after introducing this is enormous.
Just for comparison - below is a test over large rotated journals (I made a backup for the purpose of analysis as one of hosts was under attack of botnet):
```diff
  $ fail2ban-regex --version
- fail2ban-regex 0.11.2.dev3
+ fail2ban-regex 1.0.1.dev3
  $ fail2ban-regex systemd-journal[journalpath=...] sshd[mode=aggressive,logtype=journal]
  Lines: 1103027 lines, 112 ignored, 1086062 matched, 16853 missed
- [processed in 71.94 sec]
+ [processed in 28.89 sec]
```

> Hopefully there is an option to disable the feature.

So it's a plan.
But it'd be rarely needed - it's only important for filters, which:
- have multiple REs matching anything twice with 2 different REs
- and both REs shall have different meaning, e. g. one considering a failure, another (as fallback) sees the line as not a failure (some helper regex with `<F-NOFAIL>`, `<F-MLFGAINED>` etc).

Stock fail2ban filters could not conflict here. because our test-filter-factory checks for duplicate matches (and this would fail in test).

**Comment by allanwind** [comment]
On 2021-08-18T03:30:34, Sergey G. Brester wrote:
>> ... could potentially change meaning when reordered?
>
>I can hardly imagine such a filters (or rather the needs were it must be handled intentionally with different meaning for same line).

Well, you just saw me introduce such a case, and I had no idea, or 
even how to express the relevant filter when you have two events 
that are substrings of each other.  I am genuinely looking forward 
to your input on how to solve it.

>> It seems risky to mess with user input.
>
>What are the misgivings you have about this approach?

It potentially _changes_ the semantic of user input.  When
that happens silently it will surprise the when user finds 
out (much) later.  This erodes trust.

It may very well be unreasonable FUD on my part.  

The risk averse approach would be to disable the feature by 
default for upgrades (opt-in), and enable it for new installs 
(opt-out).

>The performance gain (and related reduction of fail2ban's load) after introducing this is enormous.
>Just for comparison - below is a test over large rotated journals (I made a backup for the purpose of analysis as one of hosts was under attack of botnet):
>```diff
>  $ fail2ban-regex --version
>- fail2ban-regex 0.11.2.dev3
>+ fail2ban-regex 1.0.1.dev3
>  $ fail2ban-regex systemd-journal[journalpath=...] sshd[mode=aggressive,logtype=journal]
>  Lines: 1103027 lines, 112 ignored, 1086062 matched, 16853 missed
>- [processed in 71.94 sec]
>+ [processed in 28.89 sec]
>```

That is really impressive!

>> Hopefully there is an option to disable the feature.
>
>So it's a plan.
>But it'd be rarely needed - it's only important for filters, which:
>- have multiple REs matching anything twice with 2 different REs
>- and both REs shall have different meaning, e. g. one considering a failure, another (as fallback) sees the line as not a failure (some helper regex with `<F-NOFAIL>`, `<F-MLFGAINED>` etc).

sshd, I suspect, is a really important jail.  It is sad if that 
means not being able to benefit from the performance improvement.  

Maybe failregex is too overloaded?  It is used to (1) identify
failures, (2) end sessions (F-MLFFORGET), (3) find <HOST> within a 
session (F-NOFAIL).

>Stock fail2ban filters could not conflict here. because our test-filter-factory checks for duplicate matches (and this would fail in test).

Nice!

**Comment by allanwind** [comment]
Hi Sergey,

I got another variation of this event for you:

2021-08-20T13:33:04.674+00:00 pawan sshd[1328701]: error: kex protocol error: type 30 seq 1 [preauth]

and it was banned with but I don't understand why:

# egrep '1328701|183.136.226.2' /var/log/syslog.ng > 1.log
# cat 1.log
```
2021-08-20T13:33:04.174+00:00 pawan sshd[1328701]: Connection from 183.136.226.2 port 29709 on 71.126.254.78 port 22 rdomain ""
2021-08-20T13:33:04.674+00:00 pawan sshd[1328701]: error: kex protocol error: type 30 seq 1 [preauth]
2021-08-20T13:33:23.924+00:00 pawan sshd[1328701]: Connection closed by 183.136.226.2 port 29709 [preauth]
2021-08-20T13:33:24.424+00:00 pawan fail2ban.filter[1183]: INFO [sshd] Found 183.136.226.2 - 2021-08-20 13:33:23
2021-08-20T13:33:24.924+00:00 pawan fail2ban.actions[1183]: NOTICE [sshd] Ban 183.136.226.2
```

# fail2ban-regex 1.log 'sshd[mode=aggressive]'
...
Lines: 5 lines, 2 ignored, 0 matched, 3 missed

Btw, if I removed the known prefix you mentioned earlier I would 
get a configparser.InterpolationDepthError.  No changes to sshd.conf
and here is my sshd.local:

[Definition]
mdre-ddos =
   %(known/mdre-ddos)s
   ^kex_exchange_identification: read: Connection reset by peer$
mdre-ddos-other =
   %(known/mdre-ddos-other)s
   ^<F-MLFFORGET>Connection reset</F-MLFFORGET> by <HOST>(?:%(known/__suff)s|\s*)$
   <mdre-normal-other>
failregex =
   %(known/failregex)s
   ^Received disconnect from <HOST> port .+: No more user authentication methods available. \[preauth\]$

**Comment by allanwind** [comment]
On 2021-08-20T12:45:44, Allan Wind wrote:
>I got another variation of this event for you:
>
>2021-08-20T13:33:04.674+00:00 pawan sshd[1328701]: error: kex protocol error: type 30 seq 1 [preauth]

... and another that should be banned:

```
2021-08-20T19:55:18.424+00:00 pawan sshd[1392946]: error: kex_exchange_identification: banner line contains invalid characters
```

**Comment by sebres** [comment]
> I got another variation ... and it was banned with but I don't understand why

In `aggressive` mode fail2ban banning `Connection closed ... [preauth]`, which was the result of an attempt with `kex protocol error`, so the second rule is not necessary, moreover would bother probably because this attempt would count twice (both as failures), so it'd reach `maxretry` earlier.

> ... and another that should be banned

Well, it depends which additional messages sshd wrote with same session-ID (`sshd[1392946]`)... It is not correct to consider single log-message only. Better provide whole excerpt with all messages of the session from connect to close/disconnect.

Basically many messages are there due to LogLevel VERBOSE, but even without it sshd filter is able to find all remaining messages in aggressive mode. Better (or more correct either) would be probably to provide different sets of REs for VERBOSE and non VERBOSE logging of sshd.

**Comment by allanwind** [comment]
On 2021-08-23T02:05:37, Sergey G. Brester wrote:
>> I got another variation ... and it was banned with but I don't understand why
>
>In `aggressive` mode fail2ban banning `Connection closed ... 
>[preauth]`, which was the result of an attempt with `kex protocol 
>error`, so the second rule is not necessary, moreover would 
>bother probably because this attempt would count twice (both as 
>failures), so it'd reach `maxretry` earlier.
>
>Well, it depends which additional messages sshd wrote with same session-ID (`sshd[1392946]`)... It is not correct to consider single log-message only. Better provide whole excerpt with all messages of the session from connect to close/disconnect.

I use maxretry = 1, and included all the relevant log events (per 
grep included) for both sshd pid (1328701) and ip (183.136.226.2).  
Why was it banned when fail2ban-regex showed 0 matched?

>Basically many messages are there due to LogLevel VERBOSE, but even without it sshd filter is able to find all remaining messages in aggressive mode. Better (or more correct either) would be probably to provide different sets of REs for VERBOSE and non VERBOSE logging of sshd.

Yeah, I concluded the same and subsequently disabled VERBOSE after 
I tweaked rules as we discussed it.


/Allan
-- 
Allan Wind
Yaxto - Runs Your Business
<https://yaxto.com>

**Comment by sebres** [comment]
> Why was it banned when fail2ban-regex showed 0 matched?

Because fail2ban-regex tries every regex (unless `--no-check-all` specified) and does not expect that a line would found with 2 different REs (with different meanings). 
I guess you have extended the filter (like in https://github.com/fail2ban/fail2ban/issues/3086#issuecomment-900242625), in this case fail2ban-regex find 2 matches for last line:
```
|  20) [1] ^<F-MLFFORGET>(Connection (?:closed|reset)|Disconnected)</F-MLFFORGET> (?:by|from)(?: (?:invalid|authenticating) user <F-USER>\S+|.*?</F-USER>)? <HOST>(?: (?:port \d+|on \S+)){0,2}\s+\[preauth\]\s*$
|  21) [1] ^<F-NOFAIL><F-MLFFORGET>(Connection closed|Disconnected)</F-MLFFORGET></F-NOFAIL> (?:by|from)(?: (?:invalid|authenticating) user <F-USER>\S+|.*?</F-USER>)? <HOST>(?:(?: (?:port \d+|on \S+|\[preauth\])){0,3}\s*|\s*)$
```
where 20) is matching it as failure, but 21) matching it as no-fail (and the last wins here, so it is counted as ignored).

Using `--no-check-all` option or commenting out the additional ("double" NO-FAIL) expression will match it as failure only and show:
```
3 lines, 1 ignored, 1 matched, 1 missed
```

**Comment by allanwind** [comment]
On 2021-08-24T04:06:59, Sergey G. Brester wrote:
>Using `--no-check-all` option or commenting out the additional ("double" NO-FAIL) expression will match it as failure only and show:
>```
>3 lines, 1 ignored, 1 matched, 1 missed
>```

That was what I was missing, thanks.  Should that be the default?  


/Allan
-- 
Allan Wind
Yaxto - Runs Your Business
<https://yaxto.com>

**Comment by sebres** [comment]
fixed in bf689c27b833d4cafc6ce34ada214cd4df2d7d86 for 0.10, will be merged in 0.11 and 1.0

**Comment by sysoppl** [comment]
I'm using Fail2Ban v0.11.2 and ``` error: kex_exchange_identification: Connection closed by remote host ``` with aggressive mode is not blocked.

Config:
```
[sshd]
backend = systemd
enabled = true
mode = aggressive
port = 0:65535
filter = sshd
logpath = /var/log/auth.log
maxretry = 1
bantime = -1
```
Anything I'm doing wrong?

**Comment by sebres** [comment]
Version 0.11.2 was released before this got implemented in bf689c27b833d4cafc6ce34ada214cd4df2d7d86.
You have to update [sshd-filter](https://raw.githubusercontent.com/fail2ban/fail2ban/0.11/config/filter.d/sshd.conf) manually (it should be completely compatible to 0.11.2, but needs to be tested).
Or simply add the regex to your `jail.local`:
```ini
[sshd]
...
failregex = %(known/failregex)s
            ^kex_exchange_identification: Connection closed by remote host
```
